File: generate_fit_and_write_example.cpp
#include <iostream>
#include <memory>
#include <TFile.h>
#include <TTree.h>
#include <TH1F.h>
#include <TRandom3.h>
#include <TCanvas.h>
#include <TStyle.h>
#include <TFitResult.h>
#include <TF1.h>
#include <TPaveText.h>
#include <TString.h>

// clang++ -O3 generate_fit_and_write_example.cpp -o generate_fit_and_write_example `root-config --cflags --libs`

void generate_fit_and_save() 
{
    // 1. Configuration
    TString file_name = "gaussian_data.root";
    TString output_plot_name = "gaussian_fit_results";
    int num_events = 10000;
    double mean_gen = 5.0;      // Generated mean
    double sigma_gen = 1.5;     // Generated standard deviation

    // 2. Create ROOT File
    std::unique_ptr<TFile> output_file = std::make_unique<TFile>(file_name, "RECREATE");
    std::cout << "Created file: " << file_name << std::endl;

    // 3. Create TTree and a branch to hold the data
    std::unique_ptr<TTree> tree = std::make_unique<TTree>("tree", "Tree containing generated Gaussian data");
    tree->SetDirectory(nullptr);
    double val_buffer = 0.0;
    tree->Branch("gauss_val", &val_buffer, "gauss_val/D");

    // 4. Create Histogram
    std::unique_ptr<TH1F> hist = std::make_unique<TH1F>("h_gauss", "Gaussian Distribution & Fit;Value;Events", 100, 0, 10);
    hist->SetDirectory(nullptr);

    // 5. Generate Data, Fill Histogram, and Fill Tree
    TRandom3 rand(42); // Seed 42 for reproducibility
    std::cout << "Generating " << num_events << " events..." << std::endl;
    for (int i = 0; i < num_events; ++i) {
        val_buffer = rand.Gaus(mean_gen, sigma_gen);
        hist->Fill(val_buffer);
        tree->Fill();
    }

    // 6. Fit the Histogram using ROOT's Fit Engine
    std::cout << "\n--- Fitting Histogram ---" << std::endl;
    TFitResultPtr fit_result = hist->Fit("gaus", "S");

    // Extract Fit Results
    double chi2 = fit_result->Chi2();
    int ndof = fit_result->Ndf();
    double chi2_ndof = (ndof > 0) ? chi2 / ndof : 0;
    double fit_mean = fit_result->Parameter(1);
    double fit_mean_err = fit_result->ParError(1);
    double fit_sigma = fit_result->Parameter(2);
    double fit_sigma_err = fit_result->ParError(2);

    std::cout << "\n--- Fit Results (Terminal) ---" << std::endl;
    std::cout << "Chi2 / NDF   = " << chi2 << " / " << ndof << " = " << chi2_ndof << std::endl;
    std::cout << "Fitted Mean  = " << fit_mean << " +/- " << fit_mean_err << std::endl;
    std::cout << "Fitted Sigma = " << fit_sigma << " +/- " << fit_sigma_err << std::endl;

    // 7. Create Canvas and Plot
    gStyle->SetOptStat(1110); // Show Entries, Mean, RMS
    gStyle->SetOptFit(1111);  // Show Fit probability, Chi2/ndf, errors

    std::unique_ptr<TCanvas> canvas = std::make_unique<TCanvas>("canvas", "Gaussian Fit Canvas", 800, 600);

    // Styling the histogram
    hist->SetMarkerStyle(20);
    hist->SetMarkerSize(0.8);
    hist->SetLineColor(kBlack);

    // Style the fit line (the fit function is automatically stored in hist)
    TF1* fit_func = hist->GetFunction("gaus");
    if (fit_func) {
        fit_func->SetLineColor(kRed + 1);
        fit_func->SetLineWidth(3);
    }

    // Draw with error bars (E) and standard markers (P)
    hist->Draw("EP");

    // Add custom TPaveText to show results on the canvas
    std::unique_ptr<TPaveText> pave = std::make_unique<TPaveText>(0.15, 0.70, 0.40, 0.85, "NDC");
    pave->SetBorderSize(1);
    pave->SetLineColor(kGray);
    pave->SetFillColor(0);
    pave->SetTextAlign(12); // Left-aligned
    pave->AddText(Form("#chi^{2} / ndof = %.2f / %d", chi2, ndof));
    pave->AddText(Form("Mean = %.3f #pm %.3f", fit_mean, fit_mean_err));
    pave->AddText(Form("Sigma = %.3f #pm %.3f", fit_sigma, fit_sigma_err));
    pave->Draw();

    canvas->Update();

    // 8. Save the Plots
    canvas->SaveAs(output_plot_name + ".pdf");
    std::cout << "Plot saved as " << output_plot_name << std::endl;

    // 9. Save Tree and Histogram into the ROOT file
    tree->Write();
    hist->Write();

    // Close the file
    output_file->Close();
    std::cout << "Successfully saved tree and histogram to '" << file_name << "'" << std::endl;
}

int main()
{
    generate_fit_and_save();
    return 0;
}

File: generate_fit_and_write_example.py
import array
import ROOT

def main():
    # 1. Configuration
    file_name = "gaussian_data.root"
    output_plot_name = "gaussian_fit_results"
    num_events = 10000
    mean_gen = 5.0  # Generated mean
    sigma_gen = 1.5  # Generated standard deviation

    # 2. Create ROOT File
    output_file = ROOT.TFile(file_name, "RECREATE")
    print(f"Created file: {file_name}")

    # 3. Create TTree and a branch to hold the data
    tree = ROOT.TTree("tree", "Tree containing generated Gaussian data")
    val_buffer = array.array("d", [0.0])
    tree.Branch("gauss_val", val_buffer, "gauss_val/D")

    # 4. Create Histogram
    hist = ROOT.TH1F(
        "h_gauss", "Gaussian Distribution & Fit;Value;Events", 100, 0, 10
    )

    # 5. Generate Data, Fill Histogram, and Fill Tree
    rand = ROOT.TRandom3(42)  # Seed 42
    print(f"Generating {num_events} events...")
    for _ in range(num_events):
        val = rand.Gaus(mean_gen, sigma_gen)
        hist.Fill(val)
        val_buffer[0] = val
        tree.Fill()

    # 6. Fit the Histogram using ROOT's Fit Engine
    print("\n--- Fitting Histogram ---")
    # "S" returns fit results, "Q" makes the fit quiet in terminal if desired
    fit_result = hist.Fit("gaus", "S")

    # Extract Fit Results
    chi2 = fit_result.Chi2()
    ndof = fit_result.Ndf()
    chi2_ndof = chi2 / ndof if ndof > 0 else 0
    fit_mean = fit_result.Parameter(1)
    fit_mean_err = fit_result.ParError(1)
    fit_sigma = fit_result.Parameter(2)
    fit_sigma_err = fit_result.ParError(2)

    print("\n--- Fit Results (Terminal) ---")
    print(f"Chi2 / NDF   = {chi2:.2f} / {ndof} = {chi2_ndof:.4f}")
    print(f"Fitted Mean  = {fit_mean:.4f} +/- {fit_mean_err:.4f}")
    print(f"Fitted Sigma = {fit_sigma:.4f} +/- {fit_sigma_err:.4f}")

    # 7. Create Canvas and Plot
    # Enable ROOT's elegant style settings
    ROOT.gStyle.SetOptStat(1110)  # Show Entries, Mean, RMS
    ROOT.gStyle.SetOptFit(1111)  # Show Fit probability, Chi2/ndf, errors

    canvas = ROOT.TCanvas("canvas", "Gaussian Fit Canvas", 800, 600)

    # Styling the histogram
    hist.SetMarkerStyle(20)
    hist.SetMarkerSize(0.8)
    hist.SetLineColor(ROOT.kBlack)

    # Style the fit line (the fit function is automatically stored in hist)
    fit_func = hist.GetFunction("gaus")
    if fit_func:
        fit_func.SetLineColor(ROOT.kRed + 1)
        fit_func.SetLineWidth(3)

    # Draw with error bars (E) and standard markers (P)
    hist.Draw("EP")

    # Add custom TPaveText to show cleaner results (optional decoration)
    pave = ROOT.TPaveText(0.15, 0.70, 0.40, 0.85, "NDC")
    pave.SetBorderSize(1)
    pave.SetLineColor(ROOT.kGray)
    pave.SetFillColor(0)
    pave.SetTextAlign(12)  # Left-aligned
    pave.AddText(f"#chi^{{2}} / ndof = {chi2:.2f} / {ndof}")
    pave.AddText(f"Mean = {fit_mean:.3f} #pm {fit_mean_err:.3f}")
    pave.AddText(f"Sigma = {fit_sigma:.3f} #pm {fit_sigma_err:.3f}")
    pave.Draw()

    # Update the canvas to draw everything
    canvas.Update()

    # 8. Save the Plots
    canvas.SaveAs(f"{output_plot_name}.pdf")
    print(f"Plot saved as {output_plot_name}.pdf")

    # 9. Save objects into ROOT file
    tree.Write()
    hist.Write()

    # Close the file
    output_file.Close()
    print(f"Successfully saved tree and histogram to '{file_name}'")

if __name__ == "__main__":
    main()